engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,567 @@ package ingestion
 import (
 	"context"
 	"fmt"
+	"math"
+	"strings"
 	"sync"
 
+	"github.com/coreos/go-semver/semver"
 	"github.com/rs/zerolog"
 
 	"github.com/onflow/flow-go/engine/execution/state"
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
 )
 
 // StopControl is a specialized component used by ingestion.Engine to encapsulate
-// control of pausing/stopping blocks execution.
+// control of stopping blocks execution.
 // It is intended to work tightly with the Engine, not as a general mechanism or interface.
-// StopControl follows states described in StopState
+//
+// StopControl can stop execution or crash the node at a specific block height. The stop
+// height can be set manually or by the version beacon service event. This leads to some
+// edge cases that are handled by the StopControl:
+//
+//  1. stop is already set manually and is set again manually.
+//     This is considered as an attempt to move the stop height. The resulting stop
+//     height is the new one. Note, the new height can be either lower or higher than
+//     previous value.
+//  2. stop is already set manually and is set by the version beacon.
+//     The resulting stop height is the lower one.
+//  3. stop is already set by the version beacon and is set manually.
+//     The resulting stop height is the lower one.
+//  4. stop is already set by the version beacon and is set by the version beacon.
+//     This means version boundaries were edited. The resulting stop
+//     height is the new one.
 type StopControl struct {
```

Comment:
I think it would be beneficial to move this component into its own package. For example, it defines several methods as private, partly because they should only be called by a public method which first acquires the lock. Using a separate package would let the language enforce that API contract for us.

Reply:
[b61af2a](https://github.com/onflow/flow-go/pull/4342/commits/b61af2a8773709c672ee17ccd6e777b08aa20daa)